File: chapter8_function_subgoals.py
def mean( list ):
  sum = 0
  for item in list:
    sum = sum + item

  number_of_items = len( list )
  if number_of_items == 0:
    return 0

  return sum / number_of_items


def variance( list, mean ):
  sum = 0
  for item in list:
    sum = sum + ( item - mean ) ** 2

  number_of_items = len( list )
  if number_of_items == 0:
    return 0

  return sum / number_of_items

def print_stars( list ):
  for item in list:
    item = int( item )
    print( item * '*' )

mega_collector = {}

for line in open('experimental_data.txt'):
    condition = line.split(',')[0].strip()
    measurement = float( line.split(',')[1].strip() )

    if condition not in mega_collector:
      mega_collector[ condition ] = []

    mega_collector[ condition ].append( measurement )

for condition, measurements in mega_collector.items():
  m = mean( measurements )
  v = variance( measurements , m )
  list = []
  list.append( m )
  list.append( v )
  print_stars( list )

File: chapter8_function_subgoals.r
mean <- function( list ){
  sum <- 0
  for( item in list ){
    sum <- sum + item
  }

  number_of_items <- length( list )
  if( number_of_items == 0 ){
    return(0)
  }

  return(sum / number_of_items)
}

variance <- function( list, mean ){
  sum <- 0
  for( item in list ){
    sum <- sum + ( item - mean ) ** 2
  }

  number_of_items <- length( list )
  if( number_of_items == 0 ){
    return(0)
  }

  return( sum / number_of_items )
}

print_stars <- function( list ){
  for( item in list ) {
    item = as.integer( item )
    print( paste( rep( '*', item ) , collapse = "" ) )
  }
}

mega_collector <- list()

data <- read.csv('experimental_data.txt')

for( i in 1:nrow(data) ){
  condition <- data[i, 1]
  measurement <- as.numeric( data[i, 2] )

  if( ! condition %in% names( mega_collector ) ){
    mega_collector[[ condition ]] <- list()
  }

  previous_values <- mega_collector[[ condition ]]
  new_values <- c( previous_values , measurement  )
  mega_collector[[ condition ]] <- new_values

}

for( condition in names(mega_collector) ) {
  measurements <- mega_collector[[ condition ]]
  measurements <- unlist( measurements ) ## transforming them to a simpler list
  m <- mean( measurements )
  v <- variance( measurements , m )
  l <- list()
  l <- c( l, m, v )
  print_stars( l )
}
